### PR TITLE
Mnt/jmcguig

### PR DESCRIPTION
**Pull request by jmcguigan10**

### CODE REVIEW

File: model_training/ml_neuralnet.py [hunk 1]
```diff
@@ -252,7 +253,13 @@ def predict_all(self, F4_initial):
 
         return F4_final, growthrate, stability
 
-    
+    @torch.jit.export
+    def predict_stability(self, F4_initial):
```
Comment (jmcguigan10) [inline model_training/ml_neuralnet.py:257]: Function to predict stability only. Use sigmoid to output logit.

File: model_training/ml_neuralnet.py [hunk 2]
```diff
@@ -28,6 +28,9 @@ def __init__(self, parms):
 
         # store input arguments
         self.NF = parms["NF"]
+        pair_count = 2 * self.NF
+        self.register_buffer("pair_indices", torch.triu_indices(pair_count, pair_count))
+        self.register_buffer("minkowski_metric", torch.tensor([1.0, 1.0, 1.0, -1.0]))
```
Comment (jmcguigan10) [inline model_training/ml_neuralnet.py:33]: Define minkowski tensor and dot product tensor for easier calculation later.

File: model_training/ml_neuralnet.py [hunk 3]
```diff
@@ -121,37 +124,35 @@ def X_from_F4(self, F4):
         Returns:
             torch.Tensor: Neural network input tensor. Indexed as [sim, iX]
         """
-        index = 0
         nsims = F4.shape[0]
-        X = torch.zeros((nsims, self.NX), device=F4.device)
-        F4_flat = F4.reshape((nsims, 4, 2*self.NF)) # [simulationIndex, xyzt, species]
+        if nsims == 0:
+            return torch.zeros((0, self.NX), device=F4.device, dtype=F4.dtype)
```
Comment (jmcguigan10) [inline model_training/ml_neuralnet.py:129]: Changed the manual index counter and per-pair loops; nsims is read, and if it’s zero we early-return an empty tensor with the right shape/device/dtype.

File: model_training/ml_neuralnet.py [hunk 4]
```diff
@@ -121,37 +124,35 @@ def X_from_F4(self, F4):
...
-        for a in range(2*self.NF):
-            for b in range(a,2*self.NF):
-                F1 = F4_flat[:,:,a]
-                F2 = F4_flat[:,:,b]
+        metric = self.minkowski_metric.view(1, 4, 1).to(
+            device=F4.device, dtype=F4.dtype
+        )
+        dot_matrix = torch.einsum(
+            "bma,bmc->bac", F4_flat * metric, F4_flat
+        )  # [sim, species, species]
```
Comment (jmcguigan10) [inline model_training/ml_neuralnet.py:145]: Instead of looping and calling dot4, we build a Minkowski metric tensor (self.minkowski_metric) on the correct device/dtype, compute all pairwise dot products in one einsum, and then slice the upper triangle using the precomputed pair_indices buffer. This was just an idea I had that seems to work. I'm indifferent to whether we keep it.

File: model_training/ml_trainmodel.py [hunk 5]
```diff
@@ -20,27 +20,39 @@
...
-    weights = []
-    for dataset in dataset_train_list:
-        nsamples = len(dataset)
-        weights.extend([1.0 / nsamples] * nsamples)
+    weights = torch.cat(
+        [
+            torch.full((len(dataset),), 1.0 / len(dataset), dtype=torch.double)
+            for dataset in dataset_train_list
+        ]
+    )
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:31]: Skip defining a python list and just define the torch tense. Saves overhead. Probably noticeable on large datasets according to PyTorch documentation. Not really sure why it's faster though.

File: model_training/ml_trainmodel.py [hunk 6]
```diff
@@ -20,27 +20,39 @@
...
+        weights=weights, num_samples=num_samples, replacement=True, generator=generator
+    )
+    num_workers = parms.get("num_workers", 0)
+    loader = DataLoader(
+        dataset_train,
+        batch_size=parms["batch_size"],
+        sampler=sampler,
+        num_workers=num_workers,
+        pin_memory=parms.get("pin_memory", parms.get("device", "cpu") == "cuda"),
+        persistent_workers=num_workers > 0,
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:50]: Data sampling determinism. Sets a random seed and pins everything to memory, indexed by the seed.

File: model_training/ml_trainmodel.py [hunk 7]
```diff
@@ -189,7 +204,7 @@ def train_asymptotic_model(
             F4f_pred_train, growthrate_pred_train, _ = model.predict_all(
                 F4i_asymptotic_train
             )
-            _, _, stable_pred_train = model.predict_all(F4i_stable_train)
+            stable_pred_train = model.predict_stability(F4i_stable_train)
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:207]: use the predict_stability function we defined in ml_neuralnet. This should be much faster inference.

File: model_training/ml_trainmodel.py [hunk 8]
```diff
@@ -334,16 +349,12 @@ def pcgrad_step(losses, model, optimizer):
...
             )
-            _, _, stable_pred_test = model.predict_all(F4i_stable_test)
+            stable_pred_test = model.predict_stability(F4i_stable_test)
             ndens_pred_test, fluxmag_pred_test, Fhat_pred_test = get_ndens_fluxmag_fhat(
                 F4f_pred_test
             )
-            ndens_true_test, fluxmag_true_test, Fhat_true_test = get_ndens_fluxmag_fhat(
-                F4f_true_test
-            )
-            ntot_test = ntotal(F4i_asymptotic_test)
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:344]: ignore this I switched it's location to above.

**Review by jmcguigan10 — COMMENTED** [review]
ml_neuralnet and ml_trainmodel optimization.
